horovod/run/run.py

```diff
@@ -0,0 +1,472 @@
+        for task in tasks:
+            task.notify_initial_registration_complete()
+        if verbose:
+            print("Notified all the hosts that the registration is complete.")
+        # Each worker should probe the interfaces of the next worker in a ring
+        # manner and filter only the routed ones -- it should filter out
+        # interfaces that are not really connected to any external networks
+        # such as lo0 with address 127.0.0.1.
+        if verbose:
+            print("Waiting for hosts to perform host-to-host "
+                  "interface checking.")
+        driver.wait_for_task_to_task_address_updates(tmout)
+        if verbose:
+            print("Host-to-host interface checking successful.")
+        # Determine a set of common interfaces for task-to-task communication.
+        common_intfs = set(driver.task_addresses_for_tasks(0).keys())
+        for index in range(1, num_hosts):
+            common_intfs.intersection_update(
+                driver.task_addresses_for_tasks(index).keys())
+        if not common_intfs:
+            raise Exception(
+                'Unable to find a set of common task-to-task communication '
+                'interfaces: %s'
+                % [(index, driver.task_addresses_for_tasks(index))
+                   for index in range(num_hosts)])
+        return common_intfs
+    finally:
+        driver.shutdown()
+
+
+def _is_open_mpi_installed():
+    output = six.StringIO()
+    command = 'mpirun --version'
+    try:
+        exit_code = safe_shell_exec.execute(command, stdout=output,
+                                            stderr=output)
+        output_msg = output.getvalue()
+    except Exception:
+        print(traceback.format_exc(), file=sys.stderr)
+        return False
+    finally:
+        output.close()
+
+    if exit_code == 0:
+        if 'Open MPI' not in output_msg:
+            print('Open MPI not found in output of mpirun --version.',
+                  file=sys.stderr)
+            return False
+        else:
+            return True
+    else:
+        print("Was not able to run %s:\n%s" % (command, output_msg),
+              file=sys.stderr)
+        return False
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Horovod Runner')
+
+    parser.add_argument('-v', '--version', action="store_true", dest="version",
+                        help="Show Horovod version.")
+
+    parser.add_argument('-np', '--num-proc', action="store", dest="np",
+                        type=int,
+                        help="Number of processes which should be equal to "
+                             "the total number of available GPUs.")
+
+    parser.add_argument('-p', '--ssh-port', action="store", dest="ssh_port",
+                        type=int,
+                        help="SSH port on all the workers.")
+
+    parser.add_argument('-H', '--host', action="store", dest="host",
```

Comment:
Does this imply that the user doesn't specify the number of processes to run per host as they would in `mpirun`?  In other words, users cannot launch non-homogenous jobs (jobs with differing numbers of GPUs per host)?

Reply:
the host parameter is just similar to mpirun -H arg where they can specify the number of processes per each host e.g.: `host-1:2,host-2:2,host-3:4`

Reply:
We should assume user never heard of `mpirun`, so we need to convey that he needs to specify number of processes per host as well.